packages/player/src/components/slider.tsx

```diff
@@ -0,0 +1,206 @@
+    (event: Event | SyntheticEvent) => {
+      if (onMouseMove) {
+        onMouseMove(event)
+      }
+    },
+    [onMouseMove],
+  )
+
+  const handleMouseUp = React.useCallback(
+    (event: Event) => {
+      // On iOS safari, a subsequent mouseup event will be fired after touchend.
+      // Its weird event positions make the player seek a wrong time.
+      // calling preventDefault (at touchend phase) will prevent the mouseup event
+      event.preventDefault()
+
+      document.removeEventListener('mousemove', handleMouseMove, true)
+      document.removeEventListener('mouseup', handleMouseUp, true)
+      document.removeEventListener('touchmove', handleMouseMove, true)
+      document.removeEventListener('touchend', handleMouseUp, true)
+
+      setActive(false)
+
+      if (sliderInactive) {
+        sliderInactive(event)
+      }
+
+      if (onMouseUp) {
+        onMouseUp(event)
+      }
+    },
+    [onMouseUp, sliderInactive, handleMouseMove],
+  )
+
+  function handleFocus(e: Event | SyntheticEvent) {
+    document.addEventListener('keydown', handleKeyPress, true)
+    if (props.onFocus) {
+      props.onFocus(e)
+    }
+  }
+
+  function handleBlur(e: Event | SyntheticEvent) {
+    document.removeEventListener('keydown', handleKeyPress, true)
+    if (props.onBlur) {
+      props.onBlur(e)
+    }
+  }
+
+  function handleClick(event: MouseEvent) {
+    event.preventDefault()
+    // event.stopPropagation();
+    if (props.onClick) {
+      props.onClick(event)
+    }
+  }
+
+  const handleStepForward = React.useCallback(() => {
+    if (stepForward) {
+      stepForward()
+    }
+  }, [stepForward])
+
+  const handleStepBack = React.useCallback(() => {
+    if (stepBack) {
+      stepBack()
+    }
+  }, [stepBack])
+
+  const handleKeyPress = React.useCallback(
+    (event: KeyboardEvent) => {
+      console.log(event)
+      if (event.which === 37 || event.which === 40) {
+        // Left and Down Arrows
+        event.preventDefault()
+        event.stopPropagation()
+        handleStepBack()
+      } else if (event.which === 38 || event.which === 39) {
+        // Up and Right Arrows
+        event.preventDefault()
+        event.stopPropagation()
+        handleStepForward()
+      }
+    },
+    [handleStepBack, handleStepForward],
+  )
+
+  function getProgress() {
+    const {getPercent} = props
+    if (!getPercent) {
+      return 0
+    }
+    let progress = getPercent()
+
+    // Protect against no duration and other division issues
+    if (progress < 0 || progress === Infinity) {
+      progress = 0
+    }
+    return progress
+  }
+
+  function renderChildren() {
+    const progress = getProgress()
+    const percentage = `${(progress * 100).toFixed(2)}%`
+    return React.Children.map<ReactNode, ReactNode>(props.children, (child) => {
+      if (React.isValidElement(child)) {
+        return React.cloneElement(child, {progress, percentage})
+      }
+    })
+  }
+
+  React.useEffect(() => {
+    return () => {
```

Comment:
I've never seen this before but this effect ONLY runs the cleanup function in case any of these listeners are still attached which could cause potential memory leaks